confluent_kafka/avro/serializer/__init__.py

```diff
@@ -33,8 +61,270 @@ def __str__(self):
+    :returns: The subject name with which to register the schema.
+    :rtype: str
+    """
+    if topic is None:
+        raise ValueError("Topic must be set when using TopicNameStrategy")
+
+    return "-".join([topic, '-key' if is_key else '-value'])
+
+
+def RecordNameStrategy(topic=None, is_key=False, schema=None):
+    """
+    Constructs the subject name under which a schema is registered with the Confluent Schema Registry.
+
+    RecordNameStrategy returns the fully-qualified record name regardless of the topic.
+
+    Compatibility checks of the same record name across all topics.
+    This strategy allows a topic to contain a mixture of different record types.
+
+    :param str topic: Topic name.  *Note* Not used by RecordNameStrategy
+    :param is_key: True if subject is being registered for a message key. *Note* Not used by RecordNameStrategy.
+    :param schema schema: Parsed Avro schema.
+    :raises ValueError: If schema is not set.
+    :returns: The subject name with which to register the schema.
+    :rtype: str
+    """
+    if schema is None:
+        raise ValueError("Schema must be set when using RecordNameStategy")
+
+    return schema.fullname
+
+
+def TopicRecordNameStrategy(topic=None, is_key=False, schema=None):
+    """
+    Constructs the subject name under which a schema is registered with the Confluent Schema Registry.
+
+    TopicRecordNameStrategy returns the topic name appended by the fully-qualified record name.
+
+    Compatibility checks are performed against all records of the same name within the same topic.
+    Like the RecordNameStrategy mixed record types are allowed within a topic.
+    This strategy is more flexible in that records needn't be complaint across the cluster.
+
+    :param str topic: Topic name.
+    :param schema schema: Parsed Avro schema.
+    :param is_key: True if used by a key_serializer.
+    :raises ValueError: If topic and schema are not set.
+    :returns: The subject name with which to register the schema.
+    :rtype: str
+    """
+    if not any([topic, schema]):
+        raise ValueError("Both Topic and Schema must be set when using TopicRecordNameStrategy")
+    return "-".join([topic, schema.fullname])
+
+
+class AvroSerializer(object):
+    """
+    Encodes kafka messages as Avro; registering the schema with the Confluent Schema Registry.
+
+    :param registry_client CachedSchemaRegistryClient: Instance of CachedSchemaRegistryClient.
+    :param bool is_key: True if configured as a key_serializer.
+    :param func(str, bool, schema): Returns the subject name used when registering schemas.
+    """
+
+    __slots__ = ["registry_client", "codec_cache", "is_key", "subject_strategy"]
+
+    def __init__(self, registry_client, is_key=False, subject_strategy=TopicNameStrategy):
+        self.registry_client = registry_client
+        self.codec_cache = {}
+        self.is_key = is_key
+        self.subject_strategy = subject_strategy
+
+    def __call__(self, topic, record):
```

Comment:
In C#, the serialize method is provided with everything - suggest doing the same (this was well thought through!)
https://github.com/confluentinc/confluent-kafka-dotnet/blob/1.0.x/src/Confluent.Kafka/ISerializer.cs#L47

Reply:
I considered this but instead decided to convert my serializer into a callable as you see here. I believe this adds maximum flexibility and ease of use.

Reply:
you only pass the topic and data to be serialized to the serializer. I'm suggesting you provide everything, to be maximally flexible.

Reply:
I don't think serializers care nor should they care if the contents they are serializing are a key or not. This is more a schema registry client constraint which leaked into the java clients which we are still paying for.

Reply:
Honestly the same should probably be true for `topic` but once we get to that point we have pushed serialization down to the per message level which can already be done without changes to the client.

Reply:
> I don't think serializers care nor should they care if the contents they are serializing are a key or not.

yours do!

Reply:
of course if you push all this to the message level, the user of the API would need to specify stuff twice (topic, key/value), so it's not an option, also the client can't encapsulate DRs with the original objects.

I iterated on many different strategies for this with the .net client, and the only one i couldn't fault was passing everything that could ever possibly wanted into the serdes: https://github.com/confluentinc/confluent-kafka-dotnet/blob/1.0.x/src/Confluent.Kafka/ISerializer.cs#L47

Reply:
>yours do! 

That's not technically true, the registry cares about this not the serializer. The serializer itself is agnostic to the this idea of key or value. I guess technically the serializer cares in that it throws a different type of SerializerException but as you later note this may not really be needed or desired in the first place.

Reply:
it is true - a serializer instance is specifically for either serializing keys or values.
it's both bad UX to need to specify this at construction time, and is more prone to error than what i'm suggesting.